driver-sync/src/test/functional/com/mongodb/internal/connection/OidcAuthenticationProseTests.java

```diff
@@ -256,118 +253,210 @@ public void test3p1AuthFailsWithCachedToken() throws ExecutionException, Interru
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performFind(mongoClient);
+            failCommand(391, 1, "find");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performFind(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    @Test
+    public void test4p3WriteCommandsFailIfReauthenticationFails() {
+        // Create a `MongoClient` whose OIDC callback returns one good token
+        // and then bad tokens after the first call.
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(
+                    wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performInsert(mongoClient);
+            failCommand(391, 1, "insert");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performInsert(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    private static void performInsert(final MongoClient mongoClient) {
+        mongoClient
+                .getDatabase("test")
+                .getCollection("test")
+                .insertOne(Document.parse("{ x: 1 }"));
+    }
+
+    @Test
+    public void test5p1Azure() {
+        assumeTrue(getOidcEnv().equals("azure"));
+        String oidcUri = getOidcUri();
+        assertFalse(oidcUri.contains("@"));
+        MongoClientSettings clientSettings = createSettings(oidcUri, createCallback(), null);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            // #. Perform a find operation that succeeds.
+            performFind(mongoClient);
+        }
+    }
+
+    @Test
+    public void test5p1AzureFails() {
+        assumeTrue(getOidcEnv().equals("azure"));
+        String oidcUri = getOidcUri();
+        oidcUri = oidcUri.replace("://", "://bad@");
+        MongoClientSettings clientSettings = createSettings(oidcUri, createCallback(), null);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            // #. Perform a find operation that succeeds.
+            performFind(mongoClient);
+        }
```

Comment:
I assume this test corresponds to https://github.com/mongodb/specifications/blob/master/source/auth/tests/mongodb-oidc.md#52-azure-with-bad-username. The `performFind` should fail here, yet it succeeds, and even the comment states that it is supposed to succeed.

Reply:
Fixed the test. Azure passing here https://spruce.mongodb.com/version/662c354f13c2570007349f91/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

It's a bit of an odd test, since it relies on a server error. For reference, in C# and Python:
* https://github.com/mongodb/mongo-csharp-driver/blob/8af9316816ca338a83d62d2812a3b7c06cb3c736/tests/MongoDB.Driver.Tests/Specifications/auth/OidcAuthenticationProseTests.cs#L306
* https://github.com/mongodb/mongo-python-driver/blob/4ea8fcbb3b8876923e92f4c98324a16249f00ab7/test/auth_oidc/test_auth_oidc.py#L1034